protocol/src/lib.rs

```diff
@@ -925,40 +912,263 @@ impl<'a> Handshake<'a> {
+    /// * `MaxGarbageLength`      - Buffer did not contain the garbage terminator and contains too much garbage, should not be retried.
+    fn split_garbage<'b>(&self, buffer: &'b [u8]) -> Result<(&'b [u8], &'b [u8]), Error> {
+        let garbage_term = self
+            .remote_garbage_terminator
+            .ok_or(Error::HandshakeOutOfOrder)?;
+        if let Some(index) = buffer
+            .windows(garbage_term.len())
+            .position(|window| window == garbage_term)
+        {
+            Ok((&buffer[..index], &buffer[(index + garbage_term.len())..]))
+        } else if buffer.len() >= (MAX_NUM_GARBAGE_BYTES + NUM_GARBAGE_TERMINTOR_BYTES) {
+            Err(Error::MaxGarbageLength)
+        } else {
+            // Terminator not found, the buffer needs more information.
+            Err(Error::CiphertextTooSmall)
+        }
+    }
 }
 
-/// Split off garbage in the buffer on the garbage terminator.
-///
-/// # Returns
-///
-/// A `Result` containing the garbage and the remaining ciphertext not including the terminator.
-///
-/// # Error
-///
-/// * `CiphertextTooSmall`    - Buffer did not contain a garbage terminator.
-/// * `MaxGarbageLength`      - Buffer did not contain the garbage terminator and contains too much garbage, should not be retried.
-fn split_garbage(
-    buffer: &[u8],
-    garbage_term: [u8; NUM_GARBAGE_TERMINTOR_BYTES],
-) -> Result<(&[u8], &[u8]), Error> {
-    if let Some(index) = buffer
-        .windows(garbage_term.len())
-        .position(|window| window == garbage_term)
-    {
-        Ok((&buffer[..index], &buffer[(index + garbage_term.len())..]))
-    } else if buffer.len() >= (MAX_NUM_GARBAGE_BYTES + NUM_GARBAGE_TERMINTOR_BYTES) {
-        Err(Error::MaxGarbageLength)
-    } else {
-        // Terminator not found, the buffer needs more information.
-        Err(Error::CiphertextTooSmall)
+/// High level error type for the protocol interface.
+#[cfg(feature = "std")]
+#[derive(Debug)]
+pub enum ProtocolError {
+    Io(std::io::Error),
+    Internal(Error),
+}
+
+#[cfg(feature = "std")]
+impl From<std::io::Error> for ProtocolError {
+    fn from(error: std::io::Error) -> Self {
+        ProtocolError::Io(error)
+    }
+}
+
+#[cfg(feature = "std")]
+impl From<Error> for ProtocolError {
+    fn from(error: Error) -> Self {
+        ProtocolError::Internal(error)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ProtocolError {}
+
+#[cfg(feature = "std")]
+impl fmt::Display for ProtocolError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ProtocolError::Io(e) => write!(f, "IO error: {}", e),
+            ProtocolError::Internal(e) => write!(f, "Internal error: {:?}", e),
+        }
+    }
+}
+
+/// A protocol session with handshake and send/receive packet management.
+#[cfg(feature = "async")]
+pub struct AsyncProtocol<R, W>
+where
+    R: AsyncRead + Unpin,
```

Comment:
Almost certainly you will want to bound by `Send` in both cases. `Send` allows users to move these into different threads using a `tokio::sync::Mutex`, and I can't think of an implementation where this wouldn't be `Send`

Reply:
That makes sense, added it to both the R and W type bounds.